lang/syn/src/parser/accounts/constraints.rs

```diff
@@ -740,12 +760,6 @@ impl<'ty> ConstraintGroupBuilder<'ty> {
                 "associated token mint already provided",
             ));
         }
-        if self.init.is_none() {
-            return Err(ParseError::new(
-                c.span(),
-                "init must be provided before token",
-            ));
-        }
```

Comment:
Paul, I don't understand what you mean here.
`token::mint` and others couldn't use without `init`. So there was a check on `self.init` to be `Some`.
But we are going to enable them to use without `init`.
I know you know it well. If I am doing a mistake, please let me know.

Reply:
yes but if `token` and `init` are used together, init should still go before the `token` (and other constraints).

This cant be checked inside the token checks because as you said correctly, `token::<x>` can now be used without it.

But we can move the check into the `add_init` functions. In there we should check that there is no `token::<x>` or `mint::<x>` on the builder yet, to make sure that the order remains correct, if they are used together.

Reply:
Thanks for your explanation. Everything is clear.

Reply:
But @paul-schaaf, what about this? This is currently working.
    ```
#[account(
        associated_token::mint = mint,
        associated_token::authority = payer,
        init,
        payer = payer,
    )]
    pub token: Account<'info, TokenAccount>,
```
`init` should go before the associated_token constraints?
May I write some codes to correct it?

Reply:
good catch! yes, you can